crates/holochain_cascade/tests/tests/get_record.rs

```diff
@@ -0,0 +1,307 @@
+        set_when_integrated(txn, &op_hash, Timestamp::now()).unwrap();
+    });
+    // Get should return the op.
+    let maybe_record = cascade
+        .dht_get(create_action.to_hash().into(), GetOptions::local())
+        .await
+        .unwrap();
+    assert!(maybe_record.is_some());
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn get_action_from_cache() {
+    let cache = test_cache_db();
+    let cascade = CascadeImpl::empty().with_cache(cache.clone());
+
+    let create_action = Action::Create(fixt!(Create));
+    let create_op = DhtOp::ChainOp(Box::new(ChainOp::StoreRecord(
+        fixt!(Signature),
+        create_action.clone(),
+        RecordEntry::NA,
+    )))
+    .into_hashed();
+    let op_hash = create_op.hash.clone();
+
+    cache.test_write(move |txn| {
+        insert_op_cache(txn, &create_op).unwrap();
+    });
+    // Get should not return the op while it's not valid and integrated.
+    let maybe_record = cascade
+        .dht_get(create_action.to_hash().into(), GetOptions::local())
+        .await
+        .unwrap();
+    assert!(maybe_record.is_none());
+
+    // Set op to valid.
+    let op_hash2 = op_hash.clone();
+    cache.test_write(move |txn| {
+        set_validation_status(txn, &op_hash2, ValidationStatus::Valid).unwrap();
+    });
+    // Get should not return the op while it's not integrated.
+    let maybe_record = cascade
+        .dht_get(create_action.to_hash().into(), GetOptions::local())
+        .await
+        .unwrap();
+    assert!(maybe_record.is_none());
+
+    // Set op to integrated.
+    cache.test_write(move |txn| {
+        set_when_integrated(txn, &op_hash, Timestamp::now()).unwrap();
+    });
+    // Get should return the op.
+    let maybe_record = cascade
+        .dht_get(create_action.to_hash().into(), GetOptions::local())
+        .await
+        .unwrap();
+    assert!(maybe_record.is_some());
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn get_updated_then_deleted_action() {
+    let dht = test_dht_db();
+    let cascade = CascadeImpl::empty().with_dht(dht.clone().into());
+
+    // Write a create op to the dht db.
+    let mut create = fixt!(Create);
+    create.entry_type = EntryType::App(AppEntryDef {
+        entry_index: 0.into(),
+        zome_index: 0.into(),
+        visibility: EntryVisibility::Public,
+    });
+    let create_action = Action::Create(create);
+    let create_op = DhtOp::ChainOp(Box::new(ChainOp::StoreRecord(
+        fixt!(Signature),
+        create_action.clone(),
+        RecordEntry::NotStored,
+    )))
+    .into_hashed();
+
+    let create_op2 = create_op.clone();
+    dht.test_write(move |txn| {
+        insert_op_dht(txn, &create_op2, 0, None).unwrap();
+        set_validation_status(txn, &create_op2.hash, ValidationStatus::Valid).unwrap();
+        set_when_integrated(txn, &create_op2.hash, Timestamp::now()).unwrap();
+    });
+
+    // Write an update to the dht db.
+    let mut update = fixt!(Update);
```

Comment:
What are we doing with the update in the test? I'm not seeing how this affects the next steps?

Reply:
It does not directly. I saw in the query that it filters by DHT op types StoreRecord, RegisterUpdatedRecord and RegisterDeletedBy, so I wanted to include all of these op types in the test. When the entry is deleted, the query should no longer return a record. But with an update there's actually no effect. But I can't do the query with the update action hash, because the query uses it to filter by basis.
I can remove this step or add a comment.

Reply:
This is good with a comment